ops/charm.py

```diff
@@ -1417,6 +1421,77 @@ def config(self) -> model.ConfigData:
+        config for all the options that are found in the class, but with:
+
+        * ``secret`` type options having a :class:`model.Secret` value rather
+          than the secret ID. Note that the secret object is not validated by
+          Juju at this time, so may raise :class:`SecretNotFoundError` when it
+          is later used if the secret does not exist or the unit does not have
+          permission to access it.
+        * dashes in names converted to underscores.
+
+        Any additional positional or keyword arguments will be passed through to
+        the config class.
+
+        Args:
+            cls: A class that inherits from :class:`ops.ConfigBase`.
+            errors: what to do if the config is invalid. If ``blocked``, the
+                charm will exit successfully (note that this informs Juju that
+                the event was handled and it will not be retried) after setting
+                an appropriate blocked status. If ``raise``, ``load_config``
+                will not catch any exceptions, leaving the charm to handle
+                errors.
+            args: positional arguments to pass through to the config class.
+            kwargs: keyword arguments to pass through to the config class.
+
+        Returns:
+            An instance of the config class with the current config values.
+
+        Raises:
+            ValueError: if the configuration is invalid and ``errors`` is set to
+                ``raise``.
+        """
+        config: dict[str, bool | int | float | str | model.Secret] = kwargs.copy()
+        fields = set(cls._juju_names())  # type: ignore
+        for key, value in self.config.items():
+            attr = cls._juju_name_to_attr(key)  # type: ignore
+            assert isinstance(attr, str)
+            if not attr.isidentifier():
+                if errors == 'raise':
+                    raise ValueError(f'Invalid attribute name {attr}')
+                raise model._InvalidSchemaError(status=f'Invalid attribute name {attr}')
+            if attr not in fields:
+                continue
+            option_type = self.meta.config.get(key)
+            # Convert secret IDs to secret objects. We create the object rather
+            # that using model.get_secret so that it's entirely lazy, in the
+            # same way that SecretEvent.secret is.
+            if option_type and option_type.type == 'secret' and isinstance(value, str):
+                secret = model.Secret(
+                    backend=self.model._backend,
+                    id=value,
+                    _secret_set_cache=self.model._cache._secret_set_cache,
+                )
+                config[attr] = secret
+            else:
+                config[attr] = value
+        try:
+            return cls(*args, **config)
+        except ValueError as e:
+            if errors == 'raise':
+                raise
+            raise model._InvalidSchemaError(status=f'Error in config: {e}') from e
```

Comment:
The implementation here is the one I had when we allowed charms to catch this exception for custom handling. Now that that isn't the case, the status-set could be moved here, and/or this could raise `_Abort` rather than a new exception. I lean towards keeping the flexibility of the other approach, but am happy to be convinced otherwise.

Reply:
Discussed this in person with Ben and we decided to go back to the old style.